src/libgap-api.h

```diff
@@ -1,24 +1,151 @@
 
 #ifndef LIBGAP_API_H
 #define LIBGAP_API_H
 
-#include "gap.h"
+#include "system.h"
 
-typedef void (*CallbackFunc)(void);
 
-// Initialisation and finalization
+////
+//// Setup and initialisation
+////
 
-void GAP_Initialize(int          argc,
-                    char **      argv,
-                    char **      env,
-                    CallbackFunc markBagsCallback,
-                    CallbackFunc errorCallback);
+typedef void (*GAP_CallbackFunc)(void);
+
+// TODO: document this function
+extern void GAP_Initialize(int              argc,
+                           char **          argv,
+                           char **          env,
+                           GAP_CallbackFunc markBagsCallback,
+                           GAP_CallbackFunc errorCallback);
+
+
+////
+//// program evaluation and execution
+////
+
+// Evaluate a string of GAP commands
+//
+// To see an example of how to use this function see tst/testlibgap/basic.c
+//
+// TODO: properly document this function
+extern Obj GAP_EvalString(const char * cmd);
+
+
+////
+//// variables
+////
+
+// Combines GVarName and ValGVar. For a given string, it returns the value
+// of the gvar with name <name>, or NULL if the global variable is not
+// defined.
+extern Obj GAP_ValueGlobalVariable(const char * name);
+
+
+////
+//// arithmetic
+////
+
+extern int GAP_EQ(Obj a, Obj b);
+extern int GAP_LT(Obj a, Obj b);
+extern int GAP_IN(Obj a, Obj b);
+
+extern Obj GAP_SUM(Obj a, Obj b);
+extern Obj GAP_DIFF(Obj a, Obj b);
+extern Obj GAP_PROD(Obj a, Obj b);
+extern Obj GAP_QUO(Obj a, Obj b);
+extern Obj GAP_LQUO(Obj a, Obj b);
+extern Obj GAP_POW(Obj a, Obj b);
+extern Obj GAP_COMM(Obj a, Obj b);
+extern Obj GAP_MOD(Obj a, Obj b);
+
+
+////
+//// booleans
+////
+
+extern Obj GAP_True;
+extern Obj GAP_False;
+extern Obj GAP_Fail;
+
+
+////
+//// calls
+////
+
+// Call the GAP object <func> as a function with arguments given
+// as a GAP list <args>.
+extern Obj GAP_CallFuncList(Obj func, Obj args);
+
+// Call the GAP object <func> as a function with arguments given
+// as an array <args> with <narg> entries
+extern Obj GAP_CallFuncArray(Obj func, UInt narg, Obj args[]);
+
+
+////
+//// lists
+////
+
+// Returns 1 if <obj> is a GAP list, 0 if not.
+extern int GAP_IsList(Obj obj);
+
+// Returns the length of the given GAP list.
+// If <list> is not a GAP list, an error may be raised.
+extern UInt GAP_LenList(Obj list);
+
+// Assign <val> at position <pos> into the GAP list <list>.
+// If <val> is zero, then this unbinds the list entry.
+// If <list> is not a GAP list, an error may be raised.
+extern void GAP_AssList(Obj list, UInt pos, Obj val);
+
+// Returns the element at the position <pos> in the list <list>.
+// Returns 0 if there is no entry at the given position.
+// Also returns 0 if <pos> is out of bounds, i.e., if <pos> is zero,
+// or larger than the length of the list.
+// If <list> is not a GAP list, an error may be raised.
+extern Obj GAP_ElmList(Obj list, UInt pos);
```

Comment:
Perhaps relevant here: An annoyance I've discovered is that if you try to access an element of a plist that has not actually been assigned to, then it raises a breaking error in GAP.  I don't think this should happen, at least for the API.  It should just return `NULL` or something. 

This happens when trying to process the list returned from `GAP_EvalString`.  If an expression that was evaluated returns void, then the `READ_ALL_COMMANDS` never actually assigns to the value entry in the result list

```
 182
 183                 if (evalResult) {
 184                     AssPlist(result, 2, evalResult);
 185                 }
 186
```

So if I later call something like `ELM_LIST(result, 2)` it actually results in an error.  But if I'm looking at the results of a `GAP_EvalString` the last thing I want to do is have my code jump back into the GAP interpreter which is exactly what happens here.  I can avoid this by using `ELM_PLIST` directly instead, but that isn't really desirable for the API.

Reply:
@embray this works exactly as designed and documented. DON'T USE `ELM_LIST` if yo don't want this behvarious. Use `ELM0_LIST`. And perhaps read the comments in `lists.h`, which explicit specify and document this behavior.

Well, or use `GAP_ElmList`, which does what you want, too (by wrapping `ELM0_LIST`)

Reply:
Note that (as the comment says!) the problem you describe does *not* happen with the API as implemented in this PR.

I also note that, while fixing the problem here, one'll have to do more generic error handling *some way* in the end. You can deactivate the break loop, so you'll not end up in it (`--nointeract` deactivates the break loop), and avoid this `is unbound` error, but you'll always have the potential that GAP code runs into an error, and there is currently no good way to handle this.

Reply:
I should add, it would help if the `CATCH_ERROR` macro cleared--if appropriate--the `STATE(ReadJmpError)` buffer (and that it were not jumped to if it's unset of course).  This results in errors that occur in functions like `ELM_LIST` causing longjmps to wherever the last `TRY_IF_NO_ERROR` was; often totally arbitrary.

Reply:
@fingolfin I see now that `GAP_ElmList` works as I would want it to.  I didn't see `ELM0_LIST` (ofc, it's not obvious what that should even mean).